src/io/contreaddata.c

```diff
@@ -26,25 +29,79 @@ s32 osContStartReadData(OSMesgQueue* mq) {
 void osContGetReadData(OSContPad* data) {
     u8* ptr = (u8*)__osContPifRam.ramarray;
     __OSContReadFormat readformat;
+    __OSContGCNShortPollFormat readformatgcn;
     int i;
 
-    for (i = 0; i < __osMaxControllers; i++, ptr += sizeof(__OSContReadFormat), data++) {
-        readformat = *(__OSContReadFormat*)ptr;
-        data->errno = CHNL_ERR(readformat);
+    for (i = 0; i < __osMaxControllers; i++, data++) {
+        if (__osControllerTypes[i] == CONT_TYPE_GCN) {
+            readformatgcn = *(__OSContGCNShortPollFormat*)ptr;
+            data->stick_x = ((s32)readformatgcn.stick_x) - 128;
+            data->stick_y = ((s32)readformatgcn.stick_y) - 128;
+            data->button =
+                __osTranslateGCNButtons(readformatgcn.button, readformatgcn.c_stick_x, readformatgcn.c_stick_y);
+            ptr += sizeof(__OSContGCNShortPollFormat);
+        } else {
+            readformat = *(__OSContReadFormat*)ptr;
+            data->errno = CHNL_ERR(readformat);
+
+            if (data->errno != 0) {
+                continue;
+            }
 
-        if (data->errno != 0) {
-            continue;
+            data->button = readformat.button;
+            data->stick_x = readformat.stick_x;
+            data->stick_y = readformat.stick_y;
+            ptr += sizeof(__OSContReadFormat);
         }
+    }
+}
+
+void osContGetReadDataEx(OSContPadEx* data) {
```

Comment:
The `Ex` suffix is very vague and doesn't provide much insight into what this function's purpose is. Maybe something like `Full` instead of `Ex` to better communicate that it provides all the data that can be polled from a GCN controller rather than only the N64-compatible subset?

Reply:
Are you aware of any other projects that support GCN passthough (other than HackerSM64)? I worry about creating multiple standards around the same thing, given the source of truth is Wiseguy's ultralib fork.

Reply:
I'm not aware of any projects that use it, I think it's still worthwhile to discuss a more descriptive name and change this now else it really will be set in stone

Reply:
Are there even reasons to split them out? It looks like between the implementation in HackerSM64 gamecube_controller.c and the ultralib branch, other functions got consolidated into their originals instead of being split out with the `Ex` suffix

Reply:
I'd be against merging the two together if there are any identifiable differences in function signature and/or processing of standard N64 controllers or runtime (i.e. plugging in hackerlibultra doesn't just work out of the box)

Reply:
I guess the main difference is using `OSContPadEx` instead of `OSContPad`, but from a cursory glance those might be able to just be merged too